staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go

```diff
@@ -258,8 +258,6 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatc
 				timeout = time.Duration(float64(minRequestTimeout) * (rand.Float64() + 1.0))
 			}
 			klog.V(3).InfoS("Starting watch", "path", req.URL.Path, "resourceVersion", opts.ResourceVersion, "labels", opts.LabelSelector, "fields", opts.FieldSelector, "timeout", timeout)
-			ctx, cancel := context.WithTimeout(ctx, timeout)
-			defer cancel()
```

Comment:
I'm not sure this change is actually correct.
If you would focus on the "serving part" (serveWatch) this I think isn't needed.

But we pass this ctx down to rw.Watch() [down to storage] and for that having a timeout-based context actually seems desired.,,
Why do you claim it's not needed?

also @liggitt @aojea - for another pair of eyes on this change

Reply:
I only want to remove the cancelCtx from the serving part, but rw.Watch returns the watcher associating this timeoutCtx and the watcher is used in the serving part later. So if the timeoutCtx is cancelled the serving part is aborted as well. 

But after you pointed this out, I realized this can be done by just not calling the `cancel()`, would that do?

Reply:
Not sure I'm following, but I think the existing defer cancel() is needed to actually force closure in the other direction:
- if the handler itself is closed (for whatever reason) and finishes, then the `defer cancel()` will ensure that the underlying watcher will also get closed.

That said: `defer watcher.Stop()` in serveWatch may actually already be doing that job...
https://github.com/kubernetes/kubernetes/blob/43ef1cc1972e952832c24491220ef732e11d4755/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go#L67

Reply:
I don't understand this change... why would we not call cancel here?

> Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this [Context](https://pkg.go.dev/context#Context) complete.

Reply:
This change is a preparation for https://github.com/kubernetes/kubernetes/pull/120902 so the assumption is the serveWatch will return immediately but we don't want to close the watcher, which would be done by calling cancel here.

However, I do think this is risky to not call cancel here because watcher may proactively close earlier than the timeout and leak associated with the context

Reply:
Okay, I don't think I can skip calling cancel here. The refactor I need for  https://github.com/kubernetes/kubernetes/pull/120902 can be 

```
ctx, cancel := context.WithTimeout(ctx, timeout)
watcher, err := rw.Watch(ctx, &opts)
...
# refactored from serveWatch
ws := newWatchServer(watcher, scope, outputMediaType, req, w, timeout, metrics.CleanListScope(ctx, &opts))
# mostly ServeHTTP() but it returns a func that executes code in the for {}
handle := ws.handle(w, req)

# will be executed in a separate goroutine for minimal stack memory overhead
{
  handle()
  cancel()
}
```

@wojtek-t @liggitt would this be a good cleanup/refactor to proceed?

Reply:
You may want to take a look at https://github.com/kubernetes/kubernetes/pull/122115 that I opened.

I think it might be relatively easy to get from there by simply switch `serveWatch` to return a `func()` [being then Handle* function]. So I think that evolving towards that makes sense.